api/api-environments-v3/src/main/java/com/thoughtworks/go/apiv3/environments/EnvironmentsControllerV3.java

```diff
@@ -0,0 +1,268 @@
+import static com.thoughtworks.go.api.util.HaltApiResponses.*;
+import static com.thoughtworks.go.apiv3.environments.representers.EnvironmentsRepresenter.toJSON;
+import static java.lang.String.format;
+import static java.lang.String.valueOf;
+import static java.util.Comparator.comparing;
+import static java.util.Objects.hash;
+import static java.util.stream.Collectors.toList;
+import static org.apache.commons.codec.digest.DigestUtils.md5Hex;
+import static spark.Spark.*;
+
+@Component
+public class EnvironmentsControllerV3 extends ApiController implements SparkSpringController, CrudController<EnvironmentConfig> {
+
+    private static final String SEP_CHAR = "/";
+    private final ApiAuthenticationHelper apiAuthenticationHelper;
+    private final EnvironmentConfigService environmentConfigService;
+    private final EntityHashingService entityHashingService;
+
+    @Autowired
+    public EnvironmentsControllerV3(ApiAuthenticationHelper apiAuthenticationHelper, EnvironmentConfigService environmentConfigService, EntityHashingService entityHashingService) {
+        super(ApiVersion.v3);
+        this.apiAuthenticationHelper = apiAuthenticationHelper;
+        this.environmentConfigService = environmentConfigService;
+        this.entityHashingService = entityHashingService;
+    }
+
+    @Override
+    public String controllerBasePath() {
+        return Routes.Environments.BASE;
+    }
+
+    @Override
+    public void setupRoutes() {
+        path(controllerBasePath(), () -> {
+            before("", mimeType, this::setContentType);
+            before("/*", mimeType, this::setContentType);
+
+            before("", mimeType, apiAuthenticationHelper::checkAdminUserAnd403);
+            before("/*", mimeType, apiAuthenticationHelper::checkAdminUserAnd403);
+            get("", mimeType, this::index);
+            get(Routes.Environments.NAME, mimeType, this::show);
+            post("", mimeType, this::create);
+            put(Routes.Environments.NAME, mimeType, this::update);
+            patch(Routes.Environments.NAME, mimeType, this::partialUpdate);
+            delete(Routes.Environments.NAME, mimeType, this::remove);
+
+            exception(HttpException.class, this::httpException);
+        });
+    }
+
+    public String index(Request request, Response response) throws IOException {
+        Set<EnvironmentConfig> envConfigSet = environmentConfigService.getEnvironments();
+        List<EnvironmentConfig> envViewModelList = filterUnknownAndSortEnvConfigs(envConfigSet);
+        setEtagHeader(response, calculateEtag(envViewModelList));
+        return writerForTopLevelObject(request, response, outputWriter -> toJSON(outputWriter, envViewModelList));
+    }
+
+    public String show(Request request, Response response) throws IOException {
+        String environmentName = request.params("name");
+
+        EnvironmentConfig environmentConfig = fetchEntityFromConfig(environmentName);
+
+        setEtagHeader(environmentConfig, response);
```

Comment:
This section of code has https://github.com/gocd/gocd/issues/6682 bug.
Which is fixed as part of environments API v2 https://github.com/gocd/gocd/pull/6683 PR.

Reply:
Thnks fr pointing it out. Added the fix.